File: leetcode/2001-2999/2779-maximum-beauty-of-an-array-after-applying-operation/MaximumBeautyOfAnArrayAfterApplyingOperation.ts
function maximumBeauty(nums: number[], k: number): number {
  const differences: number[] = new Array(Math.max(...nums) + 2 * k + 2).fill(0);

  nums.forEach(n => {
    differences[n]++;
    differences[n + 2 * k + 1]--;
  });

  let max = 0;
  let sum = 0;

  for (const difference of differences) {
    sum += difference;
    max = Math.max(max, sum);
  }

  return max;
}

const nums: number[] = [4, 6, 1, 2];
const k: number = 2;

console.log(maximumBeauty(nums, k));

File: leetcode/2001-3000/2701-2800/2771-2780/2779-maximum-beauty-of-an-array-after-applying-operation/MaximumBeautyOfAnArrayAfterApplyingOperation.java
import java.util.function.Function;

public class MaximumBeautyOfAnArrayAfterApplyingOperation {
  public static void main(String[] args) {
    int[] numbers = {
        4, 6, 1, 2
    };

    int k = 1;

    System.out.println((new Solution()).maximumBeauty(numbers, k));
  }
}

class Solution {
  public int maximumBeauty(int[] numbers, int k) {
    Function<int[], Integer> max = (integers) -> {
      int m = Integer.MIN_VALUE;

      for (int integer : integers) {
        m = Math.max(m, integer);
      }

      return m;
    };

    int[] differences = new int[max.apply(numbers) + 2 * k + 2];

    for (int number : numbers) {
      differences[number]++;
      differences[number + 2 * k + 1]--;
    }

    int sum = 0;
    int maxBeauty = 0;

    for (int difference : differences) {
      sum += difference;
      maxBeauty = Math.max(sum, maxBeauty);
    }

    return maxBeauty;
  }
}
